### DISCUSSION TITLE
IR package directive manipulation & plugin's debug

### DISCUSSION THREAD

**Discussion by lppedd** [ForteScarlet/kotlin-suspend-transform-compiler-plugin#31] [General]
Hey there!
Unrelated question about compiler plugins.

Ones of the gripes I have with CommonJS JS generation is the namespace wrapping.
Example: if a class' package is `com.my.company.pkg`, the class will be wrapped in a namespace `com.my.company.pkg`, while I only want
`pkg`.

I'm exploring the compiler API using this project as starting point, and I've seen how you modify an existing function to add an annotation (`JvmSynthetic`), and now I'd like to debug it, to follow what's happening line by line.

How do you begin to debug the plugin's code? Any hint?

**Comment by ForteScarlet** [comment]
Hello!
When this project was just starting out, I referred to these items:
- https://github.com/bennyhuo/kotlin-compiler-plugin-embeddable-plugin
- https://github.com/Him188/kotlin-jvm-blocking-bridge

and have tried to debug using https://github.com/tschuchortdev/kotlin-compile-testing

I currently use a rather "primitive" approach to testing: I package the plugin locally and then use a test project to compile it and observe the results.

Although I'm not currently using it, I thought you might like to look at [kotlin-compile-testing ](https://github.com/tschuchortdev/kotlin-compile-testing)

As for the mention of JS compilation, I'm probably not quite sure how that works either.

I hope this will help you with my limited abilities 😃

**Reply by lppedd** [reply]
Thanks for the pointers @ForteScarlet. Checked and starting from kotlin-compile-testing was very helpful.

I have one last question about your code.
In `SuspendTransformTransformer#postProcessGenerateOriginFunction` you're adding the annotations (e.g. `JvmSynthetic`) to the original method.
I've browsed the code but I'm failing to understand how you differentiate between JVM and JS for example inside of the transformer.

**Reply by ForteScarlet** [reply]
I have made a distinction [here](https://github.com/ForteScarlet/kotlin-suspend-transform-compiler-plugin/blob/c00c14897262a4dea87b6371695e83143d797911/compiler/suspend-transform-plugin/src/main/kotlin/love/forte/plugin/suspendtrans/symbol/SuspendTransformSyntheticResolveExtension.kt#L194) between the types of platforms.

This seems to be a work in progress at the stage of defining symbols.
